File: app/src/main/java/de/wollis_page/gibsonos/helper/DataStore.java
package de.wollis_page.gibsonos.helper;

import android.content.Context;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.util.Log;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;
import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;
import java.util.HashMap;
import java.util.Objects;

import androidx.annotation.NonNull;
import de.wollis_page.gibsonos.exception.ResponseException;
import okhttp3.MediaType;
import okhttp3.MultipartBody;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

@SuppressWarnings("unused")
public class DataStore
{
	private HashMap<String, String> params = new HashMap<>();
	
	private boolean useCache = false;
	private long cacheTTL = 2 * 60 * 1000;
	
	final String SEPERATOR = "/";
	final String DELIMITER = "/";
	final String URL_SEPERATOR = "/";
	private static final char PARAMETER_DELIMITER = '&';
	private static final char PARAMETER_EQUALS_CHAR = '=';
	
	private int timeout = 20000;

	private String token;
	private String url;
	
	private String module, task, action;
	
	private Context context;

	OkHttpClient client = new OkHttpClient();

	public DataStore(Context context, String url, String token)
	{
		this.context = context;

		if (!url.endsWith("/")) {
			url += '/';
		}

		if (!url.startsWith("http://") && !url.startsWith("https://")) {
			url = "http://" + url;
		}

		this.url = url;
		this.token = token;
	}
	
	public void addParamEncoded(String key, String value)
	{
		value = value.trim();

		try {
			value = URLEncoder.encode(value, "UTF-8");
		} catch (UnsupportedEncodingException e) {
			e.printStackTrace();
		}

		params.put(key, value);
	}
	
	public void addParam(String key, String value)
	{
		value = value.trim();
		params.put(key, value);
	}
	
	public void addParam(String key, Float value)
	{
		params.put(key, value.toString());
	}
	
	public void addParam(String key, double value)
	{
		params.put(key, Double.valueOf(value).toString());
	}
	
	public void addParam(String key, int value)
	{
		params.put(key, Integer.valueOf(value).toString());
	}
	
	public void addParam(String key, boolean value)
	{
		params.put(key, Boolean.valueOf(value).toString());
	}
	
	public void clearParams()
	{
		params.clear();
	}
	
	public void removeParam(String key)
	{
		params.remove(key);
	}
	
	public JSONObject getData() throws ResponseException {
	    if (isOnline()) {
	        return _execute();
	    }
	    	
	    return null;
	}
	
	private @NonNull RequestBody _getParams() {

		if (
			this.params == null ||
			this.params.size() == 0
		) {
			return RequestBody.create("", MediaType.get("application/json; charset=utf-8"));
		}

		MultipartBody.Builder builder = new MultipartBody.Builder().setType(MultipartBody.FORM);

		for (String key : this.params.keySet()) {
			builder.addFormDataPart(key, Objects.requireNonNull(this.params.get(key)));
		}

		return builder.build();
	}

	private JSONObject _execute() throws ResponseException {
		String url = this.getUrl();
		Log.i(Config.LOG_TAG, url);

		Request.Builder requestBuilder = new Request.Builder()
			.url(getUrl())
			.header("X-Requested-With", "XMLHttpRequest")
			.post(this._getParams())
		;

		if (this.token.length() > 0) {
			Log.i(Config.LOG_TAG, "X-Device-Token: " + this.token);
			requestBuilder.header("X-Device-Token", this.token);
		}

		try {
			Response response = client.newCall(requestBuilder.build()).execute();
			Log.i(Config.LOG_TAG, "Response code: " + response.code());
			String body = Objects.requireNonNull(response.body()).string();
			Log.i(Config.LOG_TAG, "Response body: " + body);

			JSONObject jsonResponse = new JSONObject(body);

			if (
				(jsonResponse.has("failure") && jsonResponse.getBoolean("failure")) ||
				(jsonResponse.has("success") && !jsonResponse.getBoolean("success"))
			) {
				String message = jsonResponse.has("message")
					? jsonResponse.getString("message")
					: "Fehler bei der Abfrage!"
				;

				throw new ResponseException(message, jsonResponse, response.code());
			}

			return jsonResponse;
		} catch (IOException | JSONException e) {
			e.printStackTrace();
		}
    	
    	return null;
	}
	
	private String getUrl()
	{
		String url = this.url;

		if (module != null) {
			url += module + SEPERATOR;
		}

		if (task != null) { 
			url += task + SEPERATOR;
		}

		if (action != null) {
			url += action;
		}
		
		return url;
	}
	
    private boolean isOnline()
	{
        ConnectivityManager connectivityManager = (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
        NetworkInfo activeNetworkInfo = connectivityManager != null ? connectivityManager.getActiveNetworkInfo() : null;

		return activeNetworkInfo != null && activeNetworkInfo.isConnected();
	}

	public String getModule()
	{
		return module;
	}

	public void setModule(String module)
	{
		this.module = module;
	}

	public String getTask()
	{
		return task;
	}

	public void setTask(String task)
	{
		this.task = task;
	}

	public String getAction()
	{
		return action;
	}

	public void setAction(String action)
	{
		this.action = action;
	}

	public void setRoute(String module, String task, String action)
	{
	    setModule(module);
	    setTask(task);
	    setAction(action);
    }
	
	public void setTimeout(int timeout)
	{
		this.timeout = timeout;
	}

	public boolean getCache()
	{
		return useCache;
	}

	public void setCache(boolean useCache)
	{
		this.useCache = useCache;
	}
	
	public void setCacheTimeToLive(int ttl)
	{
		cacheTTL = ttl;
	}
	
}

File: app/src/main/java/de/wollis_page/gibsonos/helper/DataStore.kt
package de.wollis_page.gibsonos.helper

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.util.Log
import de.wollis_page.gibsonos.R
import de.wollis_page.gibsonos.exception.MessageException
import de.wollis_page.gibsonos.exception.ResponseException
import de.wollis_page.gibsonos.exception.TaskException
import okhttp3.*
import okhttp3.HttpUrl.Companion.toHttpUrl
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONArray
import org.json.JSONObject
import java.io.UnsupportedEncodingException
import java.net.URLEncoder

class DataStore(url: String, private val method: String, token: String?) {
    private val params: HashMap<String, String> = HashMap()
    private val separator = "/"
    private val url: String
    private val token: String
    private val client: OkHttpClient = OkHttpClient()

    private var cacheTTL = (2 * 60 * 1000).toLong()
    private var timeout = 20000
    private var module: String? = null
    private var task: String? = null
    private var action: String? = null

    init {
        if (token == null) {
            throw TaskException("Account token doesn't exists!", R.string.account_error_no_token)
        }

        this.token = token

        var cleanUrl = url

        if (!cleanUrl.endsWith("/")) {
            cleanUrl += '/'
        }

        if (!cleanUrl.startsWith("http://") && !cleanUrl.startsWith("https://")) {
            cleanUrl = "http://$cleanUrl"
        }

        this.url = cleanUrl
    }

    fun addParamEncoded(key: String, value: String) {
        var cleanValue = value.trim { it <= ' ' }

        try {
            cleanValue = URLEncoder.encode(value, "UTF-8")
        } catch (exception: UnsupportedEncodingException) {
            exception.printStackTrace()
        }

        this.params[key] = cleanValue
    }

    fun addParam(key: String, value: String) {
        this.params[key] = value.trim { it <= ' ' }
    }

    fun addParam(key: String, value: Float) {
        this.params[key] = value.toString()
    }

    fun addParam(key: String, value: Double) {
        this.params[key] = java.lang.Double.valueOf(value).toString()
    }

    fun addParam(key: String, value: Int) {
        this.params[key] = Integer.valueOf(value).toString()
    }

    fun addParam(key: String, value: Long) {
        this.params[key] = value.toString()
    }

    fun addParam(key: String, value: Boolean) {
        this.params[key] = java.lang.Boolean.valueOf(value).toString()
    }

    fun addParam(key: String, value: List<*>) {
        this.params[key] = JSONArray(value).toString()
    }

    fun addParam(key: String, value: Map<*, *>) {
        this.params[key] = JSONObject(value).toString()
    }

    fun clearParams() {
        this.params.clear()
    }

    fun removeParam(key: String) {
        this.params.remove(key)
    }

    fun setPage(start: Long, limit: Long) {
        this.addParam("start", start)
        this.addParam("limit", limit)
    }

    private fun getParams(): RequestBody {
        if (this.params.size == 0) {
            return "".toRequestBody("application/json; charset=utf-8".toMediaType())
        }

        val builder = MultipartBody.Builder().setType(MultipartBody.FORM)

        for (key in this.params.keys) {
            Log.d(Config.LOG_TAG, "Add param '" + key + "' with value '" + this.params[key].toString() + "'")
            builder.addFormDataPart(key, this.params[key]!!)
        }

        return builder.build()
    }

    private fun getQueryParams(url: HttpUrl): HttpUrl {
        val builder = url.newBuilder()

        for (key in this.params.keys) {
            Log.d(Config.LOG_TAG, "Add param '" + key + "' with value '" + this.params[key].toString() + "'")
            builder.addQueryParameter(key, this.params[key]!!)
        }

        return builder.build()
    }

    @Throws(ResponseException::class)
    fun loadJson(): JSONObject {
        try {
            return this.checkError(this.execute())
        } catch (exception: MessageException) {
            throw exception
        } catch (exception: Exception) {
            exception.printStackTrace()

            throw ResponseException(exception.message ?: "Request has errors!", "", 0)
        }
    }

    fun loadBitmap(): Bitmap {
        try {
            return BitmapFactory.decodeStream(this.getBody(this.execute()).byteStream())
        } catch (exception: ResponseException) {
            throw exception
        } catch (exception: Exception) {
            throw ResponseException(exception.message ?: "Request has errors!", "", 0)
        }
    }

    private fun execute(): Response {
        var requestUrl = this.getUrl().toHttpUrl()
        var params: RequestBody? = null

        if (method !== "POST") {
            requestUrl = this.getQueryParams(requestUrl)
        } else {
            params = this.getParams()
        }

        Log.i(Config.LOG_TAG, "$method $requestUrl")
        val requestBuilder = Request.Builder()
            .url(requestUrl)
            .header("X-Requested-With", "XMLHttpRequest")
            .method(this.method, params)

        if (this.token.isNotEmpty()) {
            Log.i(Config.LOG_TAG, "X-Device-Token: " + this.token)
            requestBuilder.header("X-Device-Token", this.token)
        }

        val response = this.client.newCall(requestBuilder.build()).execute()
        Log.i(Config.LOG_TAG, "Response code: " + response.code)

        if (!response.isSuccessful) {
            this.checkError(response)
            throw ResponseException("Request error!", "", response.code)
        }

        return response
    }

    private fun getBody(response: Response): ResponseBody {
        return response.body ?: throw ResponseException("Body is empty!", "", response.code)
    }

    private fun getUrl(): String {
        var buildUrl = this.url

        if (!this.module.isNullOrBlank()) {
            buildUrl += this.module + this.separator
        }

        if (!this.task.isNullOrBlank()) {
            buildUrl += this.task + this.separator
        }

        if (!this.action.isNullOrBlank()) {
            buildUrl += this.action
        }

        return buildUrl
    }

    private fun checkError(response: Response): JSONObject {
        val jsonResponse = JSONObject(this.getBody(response).string())
        Log.i(Config.LOG_TAG, "Response body: $jsonResponse")

        if (
            (jsonResponse.has("failure") && jsonResponse.getBoolean("failure")) ||
            (!jsonResponse.has("success") || !jsonResponse.getBoolean("success"))
        ) {
            val message = if (jsonResponse.has("data")) {
                val data = jsonResponse.getJSONObject("data")

                when {
                    data.has("message") -> data.getString("message")
                    data.has("msg") -> data.getString("msg")
                    else -> "Response error!"
                }
            } else "Response error!"

            throw ResponseException(message, jsonResponse.toString(2), response.code)
        }

        return jsonResponse
    }

    fun setRoute(module: String?, task: String?, action: String?) {
        this.module = module
        this.task = task
        this.action = action
    }

    fun setTimeout(timeout: Int) {
        this.timeout = timeout
    }

    fun setCacheTimeToLive(ttl: Int) {
        this.cacheTTL = ttl.toLong()
    }
}